### PR TITLE
Flink: add support of writing Avro GenericRecord DataStream to Iceberg

### PR DESCRIPTION
**Pull request by stevenzwu**

### CODE REVIEW

**Comment by stevenzwu** [comment]
@hililiwei please also help take a look.

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/AvroGenericRecordConverterBase.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink;
+
+import org.junit.Test;
+
+public abstract class AvroGenericRecordConverterBase {
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/AvroGenericRecordConverterBase.java:23]: base class will be used by testing the source avro converter from RowData to Avro

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestAvroGenericRecordToRowDataMapper.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink;
+
+import org.apache.flink.table.data.RowData;
+import org.apache.iceberg.flink.AvroGenericRecordConverterBase;
+import org.apache.iceberg.flink.DataGenerator;
+import org.junit.Assert;
+
+public class TestAvroGenericRecordToRowDataMapper extends AvroGenericRecordConverterBase {
```
Comment (pvary) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestAvroGenericRecordToRowDataMapper.java:26]: Would it be worth to create a `RowDataToAvro`  test too?

**Comment by pvary** [comment]
@stevenzwu: We had several issues with Avro, RowData conversions. This is solving one of them - we might not want to solve all of them once and in one PR, but I think we should collect/consider all of them in the long run.

The specific issues:
- Iceberg shades the Avro version. This makes it inconvenient to use our own Avro tools/libraries, even if the 2 Avro version is the same. - Our current solution is to write the Avro schema to string and parse it with the shaded Avro library
- Sometimes the Iceberg schema is not exactly the same as the Avro schema:
    - Different order of the columns could cause the built in Iceberg converters to fail
    - Maybe one of the schemas could contain extra columns which are not needed after the transform
- We often need transformation in both direction:
    - When writing then Avro to RowData
    - When reading then RowData to Avro

Your PR solves one of the issues above. It might worth to consider if we can easily add a solution for some of the issues in the same PR, or we would like to solve those separately.

Thanks,
Peter

**Comment by stevenzwu** [comment]
> Iceberg shades the Avro version.

I updated the Flink doc in this PR already and pointed out that `iceberg-flink-runtime` shaded bundle jar shouldn't be used in this case, as it shade the Avro package. Instead non-shaded `iceberg-flink` jar should be used.

> Sometimes the Iceberg schema is not exactly the same as the Avro schema

This probably can be addressed separately using named mapping. Here the assumption is that the output Avro schema should be same as the Iceberg schema.

> We often need transformation in both direction:
>  - When writing then Avro to RowData
>  - When reading then RowData to Avro

This PR only addresses the write part: Avro to RowData. I will create a separate PR for the read path.

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestAvroGenericRecordToRowDataMapper.java (hunk 2 above)
Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestAvroGenericRecordToRowDataMapper.java:26]: read path will be addressed separately

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java [hunk 3]
```diff
@@ -70,6 +78,12 @@ public static class Primitives implements DataGenerator {
     private static final BigDecimal BIG_DECIMAL_NEGATIVE = new BigDecimal("-1.50");
     private static final byte[] FIXED_BYTES = "012345689012345".getBytes(StandardCharsets.UTF_8);
 
+    private static org.apache.avro.Schema addAdjustToUtc(
```
Comment (hililiwei) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java:81]: Do we need it? It doesn't seem to have been used.

Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/DataGenerators.java:81]: good catch. this is not needed anymore.

**Review by pvary — APPROVED** [review]
Thanks for all the answers!

**Review by hililiwei — APPROVED** [review]

**Comment by stevenzwu** [comment]
thanks @pvary and @hililiwei for reviewing.

**Pull request merged**